### ISSUE TITLE
data source details URL enhancements

### ISSUE THREAD

**Issue by josh-chamberlain** [Police-Data-Accessibility-Project/data-sources-app#179]
### Context

Data sources detail view

<img width="1501" alt="Screen Shot 2024-01-10 at 3 36 53 PM" src="https://github.com/Police-Data-Accessibility-Project/data-sources-app/assets/30379833/e8e6eded-44a7-4051-8725-57fa0dd219d3">

### Requirements

- [ ] `readme_url` should be clickable, like `source_url`
- [ ] display the `last_cached` property, but the label should read "Last archived"
   - we'll need a change to `middleware/data-source-queries.py` to expose that to the API
- [ ] for data sources which have a non-null `last_cached`, add a clickable secondary button reading `View archive →` under the URL, with the [fa-external-link](https://fontawesome.com/v4/icon/external-link) icon instead of the arrow.
   - the link should be structured like this. everything after the `/*/`, which is a wildcard for most recent, we just append the `source_url`. 

```
https://web.archive.org/web/*/https://www.cityofcalhoun-ga.com/wp-content/uploads/2017/12/OPEN-RECORDS-REQUEST-FORM.pdf
```

- [ ] see if we can make the line break more readily, like on `/` as well as `-` for the URL fields
   - see above for example of a bad one
   - we could also truncate to show the beginning and end of long URLs, example below
   - I'm open to other solutions, including just `overflow: hidden`

```
https://long-url.gov/4892hj89f ...
... end-of-url-j8e9gj8409vn89.pdf
```

https://web.archive.org/web/*/https://www.cityofcalhoun-ga.com/wp-content/uploads/2017/12/OPEN-RECORDS-REQUEST-FORM.pdf

**Comment by josh-chamberlain** [comment]
@mbodeantor this one would require adjustment to the endpoint, to expose `last_cached`, as well as the front end stuff. Would you prefer that be done as multiple issues/PRs, one for back end and one for front end? That's how I did it at my previous job, but I like the idea of fewer issues → a back end person handing the branch off to a front end person.

**Comment by kalenluciano** [comment]
I missed the comments about making this one single issue/branch. I did the backend component and attached the PR to this issue. Feel free to review it now or @joshuagraber can take the frontend first.

**Comment by joshuagraber** [comment]
@josh-chamberlain QQ on this before I get started:

> see if we can make the line break more readily, like on / as well as - for the URL fields

We can't do this with CSS. We can handle it with a JS utility, but I have minor concerns about this approach:
Considering the url in the screenshot (`https://www.alleghenycounty.us/Government/Departments-and-Offices/County-Council/Council-Meetings/Police-Review-Board-Meeting`), breaking on every `/` and `-` would result in something like:
```
https:/
/
www.alleghenycounty.us/
Government/
Departments-
and
...etc...
```

To know where _exactly_ to break, our utility func would need to know the runtime width of the containing grid column, the exact computed font size, and then do some math on how many characters can fit before breaking, etc... And so then we'd be getting into a lot of `DOM` querying and other Vue methods to make this work.

TLDR: We can do it, but I think it's a lot of overhead just to resolve line breaks.

For this reason, I'd advocate for just truncating text with ellipses, which we can do with 3 Tailwind classes and zero JS overhead. What do you think?

**Comment by josh-chamberlain** [comment]
@joshuagraber yeah, let's just truncate! it's nice to keep the beginning and end for disambiguation but also not totally necessary.

**Comment by joshuagraber** [comment]
Sounds good @josh-chamberlain, ty! We run into similar issues with trying to keep the end of the string and truncate in the middle, but I'll play with it and see what I come up with.

**Comment by josh-chamberlain** [comment]
@joshuagraber yeah, spending more than a few minutes on that is probably not worth it! fancy stuff for a page that will evolve, especially around how we show urls

**Comment by joshuagraber** [comment]
@josh-chamberlain Do you know any data sources offhand that have a non-null `readme_url` for testing this?

**Comment by joshuagraber** [comment]
> for data sources which have a non-null last_cached, add a clickable secondary button reading View archives → under the URL, with the [fa-external-link](https://fontawesome.com/v4/icon/external-link) icon instead of the arrow.
the link should be structured like this. everything after the /*/, which is a wildcard for most recent, we just append the source_url.

Hey @josh-chamberlain This would involve a pretty big refactor, actually. This component is looping over data and looking for items in the data source as returned from the API. I'll have to blow up the current logic in order to make this work. Do you mind if I just return `archive_link` or something from the API as well, so we don't have to add a bunch of complex logic just to render this?

**Comment by josh-chamberlain** [comment]
@joshuagraber yes, definitely finding backend ways to avoid frontend gymnastics is good. Rendering it under `last_cached` is fine by me, if it will be that easy, or we could send it via the API.

**Comment by joshuagraber** [comment]
> yes, definitely finding backend ways to avoid frontend gymnastics is good. Rendering it under last_cached is fine by me, if it will be that easy, or we could send it via the API.

@josh-chamberlain I actually just found a way around it...  Here's how it's looking currently. That is what you were thinking?

<img width="399" alt="Screenshot 2024-01-19 at 2 57 26 PM" src="https://github.com/Police-Data-Accessibility-Project/data-sources-app/assets/68428039/ddb2065c-be63-4d20-8582-e580e3ccd2cc">

This page _is_ getting pretty complex in its logic, so I may add a refactor issue to clean it up.

**Comment by josh-chamberlain** [comment]
@joshuagraber yeah, that looks good to me! closing the loop on some of these disparate software pieces is fun...

**Comment by joshuagraber** [comment]
>  @joshuagraber yeah, that looks good to me! closing the loop on some of these disparate software pieces is fun...

Agreed @josh-chamberlain! It's coming together!
